### PR TITLE
 Converting response to raise a ProblemException

### PR DESCRIPTION
**Pull request by badcure**
The goal of this to convert all possible error responses to be a raised ProblemException subclass instead. Also for flask, allow for the option to skip the error handling.

The goal is to allow the end application handle the exceptions, as well as customize the response  that aligns with the application's payload structure.

### CODE REVIEW

File: connexion/handlers.py [hunk 1]
```diff
@@ -45,12 +45,11 @@ def handle(self, *args, **kwargs):
         """
         Actual handler for the execution after authentication.
         """
-        response = problem(
+        raise AuthenticationProblem(
```
Comment (cognifloyd) [inline connexion/handlers.py]: This is causing the error in `tests/aiohttp/test_aiohttp_api_secure.py::test_auth_all_paths[pyloop]` because if no middleware handles the raised exception, then aiohttp swallows the response converting it to a 500 error:
```
>       assert get_inexistent_endpoint.status == 404
E       assert 500 == 404
E         -500
E         +404
```
You might need something like #952 to add that middleware for aiohttp.

Comment (badcure) [reply connexion/handlers.py]: Ahh okay, thanks! I will work on that today. I was confused why it was happening last night, but it was getting late for me..

It sort of makes sense now that you pointed it out.

**Comment by badcure** [comment]
Side note: In reviewing some of the errors "test_default_object_body", I plan on removing those tests.

While I have no idea how it was passing before, reviewing https://github.com/json-schema/json-schema/issues/5 seems to suggest that default should NOT be applied when something is required. Below is the exception I am seeing for this case:
```
Traceback (most recent call last):
  File "/connexion/connexion/decorators/validation.py", line 178, in validate_schema
    self.validator.validate(data)
  File "/connexion/venv/py27/lib/python2.7/site-packages/jsonschema/validators.py", line 123, in validate
    raise error
ValidationError: None is not of type 'object'

Failed validating 'type' in schema:
    {'components': {'examples': {'justAnExample': {'summary': 'a basic example.',
                                                   'value': 'Good evening, doctor.'}},
                    'parameters': {'OptionalName': {'description': 'Name of the person to greet.',
                                                    'in': 'path',
                                                    'name': 'name',
                                                    'required': False,
                                                    'schema': {'type': 'string'}}},
                    'requestBodies': {'fakeapi.hello.test_body_sanitization_body': {'content': {'application/json': {'schema': {'components': <Recursion on dict with id=140022751344632>,
                                                                                                                                'properties': {'body1': {'type': 'string'},
                                                                                                                                               'body2': {'type': 'string'}},
                                                                                                                                'type': 'object'}}},
                                                                                    'description': 'Just a testing parameter in the body',
                                                                                    'required': True}},
                    'schemas': {'new_stack': {'properties': {'image_version': {'description': 'Docker image version to deploy',
                                                                               'type': 'string'}},
                                              'required': ['image_version'],
                                              'type': 'object'}}},
     'default': {'image_version': 'default_image'},
     'properties': {'image_version': {'description': 'Docker image version to deploy',
                                      'type': 'string'}},
     'required': ['image_version'],
     'type': 'object',
     'x-body-name': 'stack',
     'x-scope': ['']}
```

File: connexion/apps/abstract.py [hunk 2]
```diff
@@ -63,8 +63,9 @@ def __init__(self, import_name, api_cls, port=None, specification_dir='',
 
         logger.debug('Specification directory: %s', self.specification_dir)
 
-        logger.debug('Setting error handlers')
-        self.set_errors_handlers()
+        if not skip_error_handlers:
+            logger.debug('Setting error handlers')
+            self.set_errors_handlers()
```
Comment (cognifloyd) [inline connexion/apps/abstract.py]: For aiohttp, the `set_error_handlers` does not actually register the exception handler. Instead the exception handling middleware is registered in `api`. So, the `skip_error_handlers` will have no effect for aiohttp. It would be nice to unify the API for Flask and Aiohttp to add the error handlers in the same way, but I don't think that fits in this PR.

File: connexion/apis/aiohttp_api.py [hunk 3]
```diff
@@ -27,7 +29,7 @@
 
 @web.middleware
 @asyncio.coroutine
-def oauth_problem_middleware(request, handler):
+def generic_problem_middleware(request, handler):
```
Comment (cognifloyd) [inline connexion/apis/aiohttp_api.py]: This will need to be rebased on master as #952 was just merged.
Now, aiohttp turns all exceptions into problems by default.

This will need to at least replace `exc.to_problem()` with `problem_from_object(exc)`.

File: connexion/__init__.py [hunk 4]
```diff
@@ -8,7 +8,6 @@
 from .exceptions import ProblemException  # NOQA
 # add operation for backwards compatability
 from .operations import compat
-from .problem import problem  # NOQA
```
Comment (cognifloyd) [inline connexion/__init__.py]: Raising ProblemExceptions is nicer and much better for everything in core connexion, but removing the ability to use `problem()` by importing directly from `connexion` is a major API change.

File: connexion/apps/abstract.py (hunk 2 above)
Comment (cognifloyd) [reply connexion/apps/abstract.py]: Maybe this should be a separate PR as well - Use this PR to convert all uses of `problem()` to `raise ProblemException()` and then another PR to add `skip_error_handlers`.
And then at some point someone can harmonize how the Flask and Aiohttp clients setup (or skip setting up) error handling.

File: connexion/problem.py [hunk 5]
```diff
@@ -40,3 +40,10 @@ def problem(status, title, detail, type=None, instance=None, headers=None, ext=N
     return ConnexionResponse(status, mimetype, content_type,
                              body=problem_response,
                              headers=headers)
+
+
+def problem_from_object(exception):
```
Comment (cognifloyd) [inline connexion/problem.py]: I'm not sure about the naming here (though it's no worse than `to_problem` which is also odd).

What is the benefit of moving this factory method from `ProblemException.to_problem` into a separate factory function?

Comment (badcure) [reply connexion/problem.py]: IMO, I have always viewed exceptions as simply exceptions. I think tracing a response through an exception(and back up to other code) can be very confusing at times. I have always found out easier to think of it as "Okay we now have this exception, let me use this other function to convert it to a response for me". 

Also, any suggestions for the name? I have no opinion on the names, I just picked something that looked similar. If you have a more preferred naming convention, I can change it.

**Comment by phillipmadlan** [comment]
Any update on this PR and possibility of merging @cognifloyd? Thanks!

**Comment by cognifloyd** [comment]
I do not work for Zalando, so I can't merge any PRs here.
But, I would love to see this included, so I've been reviewing trying to help move it along so that when the Zalando people have some time, they can just merge it.

To that end, this PR has a merge conflict since #952 was merged. I would recommend rebasing.

**Comment by ConstantinoSchillebeeckxSimpleRose** [comment]
Any movement on this, I'd love to see it merged?

**Comment by hjacobs** [comment]
Could you rebase the PR and fix the conflicts? We will take a look afterwards.

**Comment by ConstantinoSchillebeeckxSimpleRose** [comment]
[All set](https://github.com/zalando/connexion/pull/1054).

File: connexion/apis/aiohttp_api.py (hunk 3 above)
Comment (cognifloyd) [reply connexion/apis/aiohttp_api.py]: @badcure when you rebase, please also update _generic_problem which has been added to this file.

**Comment by badcure** [comment]
@ConstantinoSchillebeeckxSimpleRose and @hjacobs  Sorry for the MIA-ness. I have updated the PR with the changes. Let me know what you think

File: connexion/__init__.py (hunk 4 above)
Comment (hjacobs) [reply connexion/__init__.py]: I agree, we should not break the API without need

File: connexion/exceptions.py [hunk 6]
```diff
@@ -23,11 +21,6 @@ def __init__(self, status=400, title=None, detail=None, type=None,
         self.headers = headers
         self.ext = ext
 
-    def to_problem(self):
-        return problem(status=self.status, title=self.title, detail=self.detail,
-                       type=self.type, instance=self.instance, headers=self.headers,
-                       ext=self.ext)
-
```
Comment (cognifloyd) [inline connexion/exceptions.py]: I think we should leave this here with a deprecation warning. Removing this here is not backwards compatible.

@hjacobs What is the policy on connexion deprecations for connexion? ProblemException.to_problem was one of the primary interfaces for creating problems. It is no longer necessary with this change, but I would hate to break any code that depends on it without either a deprecation or a major version bump. Thoughts?

File: tests/test_validation.py [hunk 7]
```diff
@@ -34,40 +36,54 @@ def orig_handler(*args, **kwargs):
...
+    assert exc.value.detail == "Wrong type, expected 'integer' for path parameter 'p1'"
     request = MagicMock(path_params={'p1': '1.2'}, **kwargs)
-    assert json.loads(handler(request).data.decode())['detail'] == "Wrong type, expected 'integer' for path parameter 'p1'"
+    with pytest.raises(BadRequestProblem) as exc:
+        handler(request)
+    assert exc.value.detail == "Wrong type, expected 'integer' for path parameter 'p1'"
 
-    request = MagicMock(path_params={'p1': 1}, query={'q1': '4'}, headers={}, cookies={})
-    assert json.loads(handler(request).data.decode())['detail'].startswith('4 is greater than the maximum of 3')
+    request = MagicMock(path_params={'p1': 1}, query={'q1': '4'}, headers={})
```
Comment (cognifloyd) [inline tests/test_validation.py]: What difference does `, cookies={}` make? Why did you remove it here and in the next several changed request = lines?

File: tests/test_validation.py [hunk 8]
```diff
@@ -34,40 +36,54 @@ def orig_handler(*args, **kwargs):
...
     assert handler(request) == 'OK'
 
-    request = MagicMock(path_params={'p1': '123'}, query={}, headers={'h1': 'x'}, cookies={})
-    assert json.loads(handler(request).data.decode())['detail'].startswith("'x' is not one of ['a', 'b']")
-
-    request = MagicMock(path_params={'p1': '123'}, query={}, headers={}, cookies={'c1': 'b'})
-    assert handler(request) == 'OK'
-
-    request = MagicMock(path_params={'p1': '123'}, query={}, headers={}, cookies={'c1': 'x'})
-    assert json.loads(handler(request).data.decode())['detail'].startswith("'x' is not one of ['a', 'b']")
```
Comment (cognifloyd) [inline tests/test_validation.py]: Why did you drop these cookies tests?

File: connexion/exceptions.py (hunk 6 above)
Comment (hjacobs) [reply connexion/exceptions.py]: Yes, let's leave it here with deprecation note (and create a follow-up issue to clean up in a future release)

File: tests/test_validation.py (hunk 7 above)
Comment (badcure) [reply tests/test_validation.py]: The issue is that request.cookies was `mock.Mock` object. This was causing an issue in the underling code.

File: tests/test_validation.py (hunk 8 above)
Comment (badcure) [reply tests/test_validation.py]: My accident. Added them back. Let me know if there is anything else with the unit tests.

File: connexion/exceptions.py (hunk 6 above)
Comment (badcure) [reply connexion/exceptions.py]: Added it back with a deprecated warning. Let me know if the message should be changed.

File: connexion/__init__.py (hunk 4 above)
Comment (badcure) [reply connexion/__init__.py]: This had been added back

File: connexion/handlers.py (hunk 1 above)
Comment (badcure) [reply connexion/handlers.py]: Resolving this since everything seems to be working now. Let me know if something is off.

File: connexion/apps/abstract.py (hunk 2 above)
Comment (badcure) [reply connexion/apps/abstract.py]: Resolving since everything appears to be passing, but let me know if you feel something else should be changed.

**Comment by hjacobs** [comment]
:+1:
